Engine/Helper.cs

```diff
@@ -1910,6 +1912,37 @@ public Version GetPSVersion()
         }
 
 #endregion Methods
+
+        private struct CommandLookupKey : IEquatable<CommandLookupKey>
+        {
+            private readonly string Name;
+
+            private readonly CommandTypes CommandTypes;
+
+            internal CommandLookupKey(string name, CommandTypes? commandTypes)
+            {
+                Name = name;
+                CommandTypes = commandTypes ?? CommandTypes.All;
+            }
+
+            public bool Equals(CommandLookupKey other)
+            {
+                return CommandTypes == other.CommandTypes
+                    && Name.Equals(other.Name, StringComparison.OrdinalIgnoreCase);
+            }
+
+            public override int GetHashCode()
+            {
+                // Algorithm from https://stackoverflow.com/questions/1646807/quick-and-simple-hash-code-combinations
+                unchecked
```

Comment:
On this part, worth reading through https://github.com/PowerShell/PowerShell/issues/7134 and the comment it starts with. We discussed the hashcode stuff a bit, but not sure (1) how far we want to take the bikeshedding on hashcodes or (2) if there's a convenient function to do this in both .NET Core and .NET Framework

Reply:
PSSA already has many conditional compilation options for different version of powershell, therefore using a special, optimised version for .net core would be acceptable (the only restriction is that it should be part of .net standard 2.0 at the moment)

Reply:
Another option might be `new Tuple(Name.ToUpperInvariant(), CommandTypes).GetHashCode()`. I don't think older framework versions have `System.ValueTuple`, so this would be forced to allocate :(

Reply:
There is a NuGet package for (System.ValueTuple](https://www.nuget.org/packages/System.ValueTuple/)